### PR TITLE
[4.0] plg_system_debug - Add option to select storage method

### LINKED ISSUE #22137: [4.0] plg_system_debug doesn't log db queries to log files since pr 20380
### Steps to reproduce the issue
 - Install a nightly build where https://github.com/joomla/joomla-cms/pull/20380 is already included.
Joomla_4.0.0-alpha5-dev-Development-Update_Package.zip
- Check that your Joomla `administrator/logs/` directory is empty.
- Goto plugins > System - Debug
- Activate all possible options. Save plugin 
- Activate Debug (Global Configuration)
- Load some pages of your site
- Have a look into your Joomla `administrator/logs/` directory

### Expected result
- Log files *.sql with database queries
 

### Actual result
- No log files with database queries.

### PR DESCRIPTION
**Pull request by elkuku**
Pull Request for Issue #22137 .

### Summary of Changes

This will add an option to select a storage type for the debug plugin.

Currently only two options are supported:

1. File - Debug information will be written to disk in json format.
1. None

### Testing Instructions

1. Activate debug in global configuration
1. Switch the storage type in plugin configuration and confirm that files are written - or not.

storage-setting

### Expected result

Ability to Set the storage type to 'None'

### Actual result

Files are always written - this might be a security issue.

### Documentation Changes Required

Yes.
Add words to description of #20380 (TBD)

close #22137

### CODE REVIEW

**Comment by brianteeman** [comment]
why is this a new fieldset? Surely it belongs with logging?

**Comment by elkuku** [comment]
I think we should separate this from logging.
I believe that 'file' is just the first option and others will follow with
additional settings. So the tab will get some more content...

Brian Teeman <notifications@github.com> schrieb am Sa., 15. Sep. 2018,
03:54:

File: plugins/system/debug/debug.php [hunk 1]
```diff
@@ -600,4 +601,26 @@ private function collectLogs(): self
 
 		return $this;
 	}
+
+	/**
+	 * Setup the storage method where debug information will be store (or not).
```
Comment (Quy) [inline plugins/system/debug/debug.php]: Change to `stored`

**Comment by elkuku** [comment]
Correct.
But I think the sentence is not quite right - the "where" doesn't sound
correct. @brianteeman ?

Quy <notifications@github.com> schrieb am Sa., 15. Sep. 2018, 22:36:

**Comment by brianteeman** [comment]
Storage only refers to logging so it should be in the logging tab

**Comment by brianteeman** [comment]
It also makes no sense at all what the "storage" is referring to when it is on its own tab. 

It needs to be on the logging tab AND only shown when logging is actually enabled

**Comment by elkuku** [comment]
Back on my keyboard...
So, what I am trying to say is that

* `Logging` refers basically to "Messages" that are generated by the system during execution, deprecated messages or messages by 3PD extensions.
I believe activating this should not imply that those messages are going to be written to text files.
* `Storage`  refers to the way the whole "debug bar" data (like system info, session and request data, database queries and, well, the log messages) are handled.
&nbsp;&nbsp;Current options include:
    * `File`
    * `None`
Other possible options would be
    * `Redis`
    * `PDO`
    * A custom handler

So, in my humble opinion, this deserves two separated tabs.

On the other hand, if that doesn't make any sense to you, I would be willing, and able, to perform the required changes to get this merged so I can use my keyboard for writing code instead of comments :wink:

**Comment by brianteeman** [comment]
please make the requested changes - if the other stuff is ever written then the decision can always be reviewed at that time

**Comment by elkuku** [comment]
Done **.**

File: plugins/system/debug/debug.php (hunk 1 above)
Comment (elkuku) [reply plugins/system/debug/debug.php]: Done

**Comment by elkuku** [comment]
Partially reverted...
If you think that the storage option doesn't deserve its own tab it should go to the general "Plugin" tab, because it has absolutely no connection to the log tab for the reasons stated above.

**Comment by brianteeman** [comment]
I stick by my original comment

**Comment by mbabker** [comment]
This isn't a logging configuration in the sense of dealing with log files, so IMO it's right that it does not go onto the logging tab.

**Comment by brianteeman** [comment]
what is it then? Maybe i have completely misunderstood it but from what I understand its purpose is to decide if log files are created

**Comment by mbabker** [comment]
It's to configure whether the temporary files for a debugged request's data is stored basically, used in conjunction with the debug bar's [Open Handler](http://phpdebugbar.com/docs/openhandler.html) feature.  It basically gives you the ability to review data for previous requests.

screen shot 2018-09-18 at 7 20 07 am

**Comment by mbabker** [comment]
Well, when this option is fully integrated it wouldn't be "just" whether the data is stored but what backend storage is used (as hinted at earlier in this thread and in another issue).

**Comment by brianteeman** [comment]
Then the word storage is definitely misleading to me although now I see exactly what it is I see it is called storage in the code http://phpdebugbar.com/docs/storage.html

So if the field was called something like  "storing collected data" or "collecting data storage" that would be less confusing perhaps?

**Comment by mbabker** [comment]
"Collected Data Storage" should work.

**Comment by elkuku** [comment]
Thank you very much.
Let's blame Mrs. Hamilton, my English teacher at school... 
I also could have posted the link to the docs about storage, not sure why I omitted that - My fault.

**Review by brianteeman — APPROVED** [review]
approve language strings

**Comment by brianteeman** [comment]
Mrs Hamilton did a good job

**Comment by elkuku** [comment]
OK since now we all know what we are talking about... I'd like to pollute my own PR with something a little OT..

I just tried to implement the open handler.
The following code:
```php
# /components/com_content/Controller/DisplayController.php

public function xxx()
{
	$debugBar = new DebugBar();
	$debugBar->setStorage(new FileStorage($this->app->get('tmp_path')));

	$handler = new OpenHandler($debugBar);
	$handler->handle();
}
```
and
```php
# /plugins/system/debug/debug.php:293

$debugBarRenderer
  ->setOpenHandlerUrl(JUri::root(true) . '/index.php?option=com_content&task=xxx&format=raw');
```
will enable the browsing of previous requests inside the debug bar which, IMHO, is a *killer feature* since it will enable the average developer for the first time to examine what happens during a redirect without the requirement for a single step debugger.

Now as you might have noticed the code is "hacked" into `com_content` which is obviously not the right place.
On the other hand I think that adding a new component (`com_debug`) to add this functionality would be an overkill...

Suggestions?

**Comment by mbabker** [comment]
Add a view to com_admin.

**Comment by elkuku** [comment]
And for the front end?
A simple controller task that returns a JSON response would be enough.
Or do we need some overriding strategy here?

Michael Babker <notifications@github.com> schrieb am Di., 18. Sep. 2018,
10:32:

**Comment by elkuku** [comment]
> Add a view to com_admin.

I tried hard but I could not find a way to display anything from com_admin in frontend :( 
Any hint how to accomplish this would be very much appreciated.

**Comment by elkuku** [comment]
Not sure if we want to explore how to disable this great feature or are we good with the security enhancement that will be introduced by #22327 (which also includes the answer to my OT question here)?
So this could be closed.

**Comment by ghost** [comment]
Closed as stated above.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/22188">issues.joomla.org/tracker/joomla-cms/22188</a>.</sub>

**Pull request closed without merging**